torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2341,228 +2547,168 @@ def _load_sharded_state_dict(
+                not self._is_root
+                and handle._config.sharding_strategy == HandleShardingStrategy.FULL_SHARD
+                for handle in self._handles
+            ]
+            reshard_fn = functools.partial(
+                self._reshard,
+                self._handles,
+                free_unsharded_flat_params,
+            )
+            self._pre_forward(self._handles, unused, unshard_fn, unused, unused)
+            for handle in self._handles:
+                assert handle.flat_param.device == self.compute_device
+            output = self._fsdp_wrapped_module(*args, **kwargs)
+            return self._post_forward(self._handles, reshard_fn, unused, unused, unused, output)
 
-    @torch.no_grad()
-    def _write_back_current_shard(self, full_params):
-        """
-        Writes back full_params into self.params.
+    def _pre_forward(
+        self,
+        handles: List[FlatParamHandle],
+        reshard_fn: Optional[Callable],
+        unshard_fn: Optional[Callable],
+        module: nn.Module,
+        input: Any,
+    ):
         """
-        for p, (full_param, _) in zip(self.params, full_params):
-            if not p._is_sharded:  # type: ignore[attr-defined]
-                continue  # Already copied because no sharding.
+        Runs the pre-forward logic. This includes an opportunity to reshard
+        currently unsharded parameters such as those from previous forwards; an
+        opportunity to unshard currently sharded parameters such as those for
+        the current forward; and registering post-backward hooks for these
+        current parameters.
 
-            # TODO: Might be able to refactor to use _get_shard.
-            chunks = full_param.chunk(self.world_size)  # type: ignore[attr-defined]
-            assert len(chunks) > self.rank
-            chunk = chunks[self.rank]
-            p._local_shard.copy_(chunk)  # type: ignore[attr-defined]
+        Args:
+            handles (List[FlatParamHandle]): Handles giving the parameters
+                used in the current forward.
+            reshard_fn (Optional[Callable]): A callable to reshard any
+                currently unsharded parameters (e.g. from previous forwards) or
+                ``None`` to not do any resharding.
+            unshard_fn (Optional[Callable]): A callable to unshard any
+                currently sharded parameters or ``None`` to not do any
+                unsharding.
+            module (nn.Module): Unused; expected by the hook signature.
+            input (Any): Unused; expected by the hook signature.
+        """
+        with torch.autograd.profiler.record_function("FullyShardedDataParallel.forward"):
+            self.training_state = TrainingState_.FORWARD
+            self._exec_order_data.record_pre_forward(handles)
+            for handle in handles:
+                handle._training_state = HandleTrainingState.FORWARD
+            if reshard_fn is not None:
+                reshard_fn()
```

Comment:
Originally, for the most general non-recursive wrapping, we explored the opportunity to unshard and to reshard in _both_ the pre-forward and the post-forward.
- Pre-forward reshard: In the pre-forward of a child module, we can possibly reshard the parent module's parameters instead of waiting until the parent module's post-forward.
- Post-forward unshard: In the post-forward of a child module, we can possibly unshard the parent module's parameters instead of already doing so in the parent module's pre-forward.

However, without static graph and correct parameter-level info, we cannot leverage these two opportunities.

I think given this, I can remove the `unshard_fn` from post-forward and `reshard_fn` from pre-forward, and we can add them back later if needed.